designer/server/src/main/scala/pl/touk/nussknacker/ui/api/ExpressionSuggesterResources.scala

```diff
@@ -0,0 +1,83 @@
+package pl.touk.nussknacker.ui.api
+
+import akka.http.scaladsl.server.{Directives, Route}
+import de.heikoseeberger.akkahttpcirce.FailFastCirceSupport
+import io.circe.generic.JsonCodec
+import pl.touk.nussknacker.ui.security.api.LoggedUser
+
+import scala.collection.Map
+import scala.concurrent.ExecutionContext
+
+class ExpressionSuggesterResources(expressionSuggester: ExpressionSuggester)(implicit ec: ExecutionContext) extends Directives with FailFastCirceSupport with RouteWithUser {
+  def securedRoute(implicit user: LoggedUser): Route =
+    path("expressionSuggestions") {
+      post {
+        entity(as[ExpressionSuggestionRequest]) { expressionSuggestionRequest =>
+          complete {
+            expressionSuggester.expressionSuggestions(expressionSuggestionRequest.expression, expressionSuggestionRequest.caretPosition2d, expressionSuggestionRequest.variables)
+          }
+        }
+      }
+    }
+}
+
+class ExpressionSuggester() {
+
+  def expressionSuggestions(inputValue: String, caretPosition2d: CaretPosition2d, variables: Map[String, RefClazz]): List[ExpressionSuggestion] = {
+    val normalized = normalizeMultilineInputToSingleLine(inputValue, caretPosition2d)
+    val lastExpressionPart = focusedLastExpressionPartWithoutMethodParens(normalized.normalizedInput, normalized.normalizedCaretPosition)
+    getSuggestions(lastExpressionPart, variables)
+  }
+
+  private def normalizeMultilineInputToSingleLine(inputValue: String, caretPosition2d: CaretPosition2d): Normalized = {
+    val rows = inputValue.split("\n")
+    val trimmedRows = rows.map { row =>
+      val trimmedAtStartRow = row.dropWhile(_ == ' ').mkString("")
+      (trimmedAtStartRow, row.length - trimmedAtStartRow.length)
+    }
+    val beforeCaretInputLength = trimmedRows.take(caretPosition2d.row).map(_._1.length).sum
+    val normalizedCaretPosition = caretPosition2d.column - trimmedRows(caretPosition2d.row)._2 + beforeCaretInputLength
+    val normalizedInput = trimmedRows.map(_._1).mkString("")
+    Normalized(normalizedInput, normalizedCaretPosition)
+  }
+
+  private def getSuggestions(value: String, variables: Map[String, RefClazz]): List[ExpressionSuggestion] = {
+    val variableNotSelected = variables.keys.exists(variable => variable.toLowerCase.startsWith(value.toLowerCase))
+
+    if (variableNotSelected && value.nonEmpty) {
+      val allVariablesWithClazzRefs = variables.map { case (key, value) => ExpressionSuggestion(key, value)}.toList
+      filterSuggestionsForInput(allVariablesWithClazzRefs, value)
+    } else {
+      Nil
+    }
+  }
+
+  private def filterSuggestionsForInput(variables: List[ExpressionSuggestion], inputValue: String): List[ExpressionSuggestion] = {
+    variables.filter(variable => variable.methodName.toLowerCase.contains(inputValue.toLowerCase))
+  }
+
+  private def focusedLastExpressionPartWithoutMethodParens(expression: String, caretPosition: Int): String = {
```

Comment:
I thought that you will use `Typer` or `SpelExpressionParser`. Let's try to use one of them - see `SpelSubstitutionsCollector:27`. Thanks to that we will see if api is correct (it should be necessary to provide some processing type data)